enterprise/frontend/src/embedding-sdk/lib/license-problem.ts

```diff
@@ -0,0 +1,71 @@
+import { match } from "ts-pattern";
+import { t } from "ttag";
+
+import type { SDKConfig } from "embedding-sdk/types";
+import type { SdkLicenseProblem } from "embedding-sdk/types/license-problem";
+
+import { getIsLocalhost } from "./is-localhost";
+
+interface SdkProblemOptions {
+  config: SDKConfig;
+  hasFeatureFlag: boolean;
+}
+
+const PROBLEMS = {
+  API_KEYS_WITHOUT_LICENSE: t`The embedding SDK is using API keys. This is intended for evaluation purposes and works only on localhost. To use on other sites, a license is required and you need to implement SSO.`,
+  API_KEYS_WITH_LICENSE: t`The embedding SDK is using API keys. This is intended for evaluation purposes and works only on localhost. To use on other sites, implement SSO.`,
+  SSO_WITHOUT_LICENSE: t`Usage without a valid license for this feature is only allowed for evaluation purposes, using API keys and only on localhost. Attempting to use this in other ways is in breach of our usage policy.`,
+  CONFLICTING_AUTH_METHODS: t`You cannot use both JWT and API key authentication at the same time.`,
+} as const;
+
+export const SDK_SSO_DOCS_LINK =
+  "https://github.com/metabase/metabase/blob/master/enterprise/frontend/src/embedding-sdk/README.md#authenticate-users-from-your-back-end";
+
+/**
+ * Determine whether is a problem with user's use case of the embedding sdk.
+ * This is determined by their license and the auth method they are using.
+ */
+export function getSdkLicenseProblem(
+  options: SdkProblemOptions,
+): SdkLicenseProblem | null {
+  const {
+    hasFeatureFlag,
+    config: { jwtProviderUri, apiKey },
+  } = options;
+
+  const isSSO = !!jwtProviderUri;
+  const isApiKey = !!apiKey;
+  const isLocalhost = getIsLocalhost();
+
+  return match({ hasFeatureFlag, isSSO, isApiKey, isLocalhost })
+    .with({ isSSO: true, isApiKey: true }, () =>
+      toError(PROBLEMS.CONFLICTING_AUTH_METHODS),
+    )
+    .with({ isSSO: true, hasFeatureFlag: true }, () => null)
+    .with({ isSSO: true, hasFeatureFlag: false }, () =>
```

Comment:
This condition doesn't work for me. I think the reason is that the `SdkLicenseProblemDisplay` is nested under some components that require authentication. When the authentication fails in this case, since there's no feature flag, JWT won't work. It fails to render the error entirely.
![image](https://github.com/user-attachments/assets/09bf9ad1-42c5-468a-a89d-62b4471a2693)

Reply:
@WiNloSt I think that is likely because of the CORS error in `GET /api/session/properties` as well, as that returns a CORS error in the screenshot, we do not have the token features, so `hasFeatureFlag` becomes `true`.

(We fallback to `true` here because we don't exactly know when the settings would be populated, so we don't show a flash of error)

```
  const hasFeatureFlag = useSelector(state => {
    // When the settings endpoint has not been called, we assume that the feature is enabled.
    if (!state.settings.values?.["token-features"]) {
      return true;
    }

    // TODO: replace this with "embedding-sdk" once the token feature PR landed.
    return getTokenFeature(state, "embedding");
  });
```

Reply:
I think the logic should change then. If you can run this function when everything is ready e.g. the SDK is initialized, or `GET /api/session/properties` fails for example, then `hasFeatureFlag` should be `false` by default and is only set to `true` when we can read the settings.

having it as `true` by default makes it [not following the spec ](https://www.notion.so/metabase/Add-visual-cues-in-SDK-around-evaluation-usage-and-feature-flag-004be55a2ed247839cc9521e9a58f6d4?pvs=4#7c7e5a08cc0e4fea845163303600b11c)